File: apps/web/src/lib/api/ops.ts
"use client";

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import type { Schemas } from "./analytics";
import { api, unwrap } from "./client";

export type ReleaseSummary = Schemas["ReleaseSummary"];
export type ReleaseOut = Schemas["ReleaseOut"];
export type ReleaseCreate = Schemas["ReleaseCreate"];
export type ReleaseUpdate = Schemas["ReleaseUpdate"];
export type ReleaseStatus = Schemas["ReleaseStatus"];
export type ReleaseImpact = Schemas["ReleaseImpact"];
export type ImpactMetric = Schemas["ImpactMetric"];
export type ReleaseEventOut = Schemas["ReleaseEventOut"];
export type ChecklistOut = Schemas["ChecklistOut"];
export type ChecklistItem = Schemas["ChecklistItem"];
export type SopOut = Schemas["SopOut"];
export type SopCreate = Schemas["SopCreate"];
export type SopItem = Schemas["SopItem"];
export type KnowledgeSummary = Schemas["KnowledgeSummary"];
export type KnowledgeOut = Schemas["KnowledgeOut"];
export type KnowledgeCreate = Schemas["KnowledgeCreate"];
export type KnowledgeUpdate = Schemas["KnowledgeUpdate"];
export type FeedbackOut = Schemas["FeedbackOut"];
export type FeedbackCreate = Schemas["FeedbackCreate"];
export type FeedbackUpdate = Schemas["FeedbackUpdate"];
export type FeedbackStatus = Schemas["FeedbackStatus"];
export type FeedbackSource = Schemas["FeedbackSource"];
export type FeedbackSentiment = Schemas["FeedbackSentiment"];
export type ThemeSummary = Schemas["ThemeSummary"];
export type StakeholderRef = Schemas["StakeholderRef"];
export type DecisionSummary = Schemas["DecisionSummary"];
export type DecisionOut = Schemas["DecisionOut"];
export type DecisionCreate = Schemas["DecisionCreate"];
export type DecisionUpdate = Schemas["DecisionUpdate"];
export type DecisionStatus = Schemas["DecisionStatus"];
export type EntityRef = Schemas["EntityRef"];
export type SearchResponse = Schemas["SearchResponse"];
export type SearchHit = Schemas["SearchHit"];

// --------------------------------------------------------------------------- releases

export function useReleases(filters: { platform?: string; status?: ReleaseStatus } = {}) {
  return useQuery({
    queryKey: ["releases", filters],
    queryFn: async () => unwrap(await api.GET("/api/releases", { params: { query: filters } })),
  });
}

export function useRelease(id: number) {
  return useQuery({
    queryKey: ["releases", "one", id],
    queryFn: async () =>
      unwrap(await api.GET("/api/releases/{release_id}", { params: { path: { release_id: id } } })),
  });
}

export function useReleaseImpact(id: number, enabled = true) {
  return useQuery({
    queryKey: ["releases", "impact", id],
    queryFn: async () =>
      unwrap(await api.GET("/api/releases/{release_id}/impact", { params: { path: { release_id: id } } })),
    enabled,
    staleTime: 10 * 60_000,
  });
}

export function useReleaseMutations(id?: number) {
  const qc = useQueryClient();
  const invalidate = () => {
    qc.invalidateQueries({ queryKey: ["releases"] });
    qc.invalidateQueries({ queryKey: ["ops", "checklists"] });
  };
  const create = useMutation({
    mutationFn: async (body: ReleaseCreate) => unwrap(await api.POST("/api/releases", { body })),
    onSuccess: invalidate,
  });
  const update = useMutation({
    mutationFn: async (body: ReleaseUpdate) =>
      unwrap(await api.PATCH("/api/releases/{release_id}", { params: { path: { release_id: id! } }, body })),
    onSuccess: invalidate,
  });
  const addNote = useMutation({
    mutationFn: async (note: string) =>
      unwrap(
        await api.POST("/api/releases/{release_id}/events", {
          params: { path: { release_id: id! } },
          body: { note, kind: "note" },
        }),
      ),
    onSuccess: invalidate,
  });
  const runSop = useMutation({
    mutationFn: async (sopId: number) =>
      unwrap(
        await api.POST("/api/releases/{release_id}/checklists", {
          params: { path: { release_id: id! }, query: { sop_id: sopId } },
        }),
      ),
    onSuccess: invalidate,
  });
  const remove = useMutation({
    mutationFn: async () =>
      unwrap(await api.DELETE("/api/releases/{release_id}", { params: { path: { release_id: id! } } })),
    onSuccess: invalidate,
  });
  return { create, update, addNote, runSop, remove };
}

// --------------------------------------------------------------------------- SOPs & checklists

export function useSops() {
  return useQuery({
    queryKey: ["ops", "sops"],
    queryFn: async () => unwrap(await api.GET("/api/ops/sops")),
  });
}

export function useChecklists(filters: { status?: Schemas["ChecklistStatus"]; release_id?: number } = {}) {
  return useQuery({
    queryKey: ["ops", "checklists", filters],
    queryFn: async () => unwrap(await api.GET("/api/ops/checklists", { params: { query: filters } })),
  });
}

export function useOpsMutations() {
  const qc = useQueryClient();
  const invalidateOps = () => qc.invalidateQueries({ queryKey: ["ops"] });
  const createSop = useMutation({
    mutationFn: async (body: SopCreate) => unwrap(await api.POST("/api/ops/sops", { body })),
    onSuccess: invalidateOps,
  });
  const startChecklist = useMutation({
    mutationFn: async (body: Schemas["ChecklistCreate"]) =>
      unwrap(await api.POST("/api/ops/checklists", { body })),
    onSuccess: () => {
      invalidateOps();
      qc.invalidateQueries({ queryKey: ["releases"] });
    },
  });
  const toggleItem = useMutation({
    mutationFn: async (args: { checklistId: number; key: string; done: boolean }) =>
      unwrap(
        await api.PATCH("/api/ops/checklists/{checklist_id}/items/{key}", {
          params: { path: { checklist_id: args.checklistId, key: args.key } },
          body: { done: args.done },
        }),
      ),
    onSuccess: () => {
      invalidateOps();
      qc.invalidateQueries({ queryKey: ["releases"] });
    },
  });
  const createDoc = useMutation({
    mutationFn: async (body: KnowledgeCreate) => unwrap(await api.POST("/api/ops/knowledge", { body })),
    onSuccess: invalidateOps,
  });
  const updateDoc = useMutation({
    mutationFn: async (args: { id: number; body: KnowledgeUpdate }) =>
      unwrap(
        await api.PATCH("/api/ops/knowledge/{doc_id}", {
          params: { path: { doc_id: args.id } },
          body: args.body,
        }),
      ),
    onSuccess: invalidateOps,
  });
  const removeDoc = useMutation({
    mutationFn: async (id: number) =>
      unwrap(await api.DELETE("/api/ops/knowledge/{doc_id}", { params: { path: { doc_id: id } } })),
    onSuccess: invalidateOps,
  });
  const createFeedback = useMutation({
    mutationFn: async (body: FeedbackCreate) => unwrap(await api.POST("/api/ops/feedback", { body })),
    onSuccess: invalidateOps,
  });
  const updateFeedback = useMutation({
    mutationFn: async (args: { id: number; body: FeedbackUpdate }) =>
      unwrap(
        await api.PATCH("/api/ops/feedback/{feedback_id}", {
          params: { path: { feedback_id: args.id } },
          body: args.body,
        }),
      ),
    onSuccess: invalidateOps,
  });
  return {
    createSop,
    startChecklist,
    toggleItem,
    createDoc,
    updateDoc,
    removeDoc,
    createFeedback,
    updateFeedback,
  };
}

// --------------------------------------------------------------------------- knowledge

export function useKnowledge(filters: { tag?: string; q?: string } = {}) {
  return useQuery({
    queryKey: ["ops", "knowledge", filters],
    queryFn: async () =>
      unwrap(
        await api.GET("/api/ops/knowledge", {
          params: { query: { ...filters, q: filters.q && filters.q.length >= 2 ? filters.q : undefined } },
        }),
      ),
  });
}

export function useKnowledgeDoc(id: number | null) {
  return useQuery({
    queryKey: ["ops", "knowledge", "one", id],
    queryFn: async () =>
      unwrap(await api.GET("/api/ops/knowledge/{doc_id}", { params: { path: { doc_id: id! } } })),
    enabled: id !== null,
  });
}

// --------------------------------------------------------------------------- feedback

export function useFeedback(
  filters: {
    status?: FeedbackStatus;
    theme?: string;
    open_only?: boolean;
    linked_type?: string;
    linked_id?: number;
  } = {},
) {
  return useQuery({
    queryKey: ["ops", "feedback", filters],
    queryFn: async () => unwrap(await api.GET("/api/ops/feedback", { params: { query: filters } })),
  });
}

export function useFeedbackThemes() {
  return useQuery({
    queryKey: ["ops", "feedback", "themes"],
    queryFn: async () => unwrap(await api.GET("/api/ops/feedback/themes")),
  });
}

export function useStakeholders() {
  return useQuery({
    queryKey: ["ops", "stakeholders"],
    queryFn: async () => unwrap(await api.GET("/api/ops/feedback/stakeholders")),
    staleTime: 10 * 60_000,
  });
}

// --------------------------------------------------------------------------- decisions

export function useDecisions(
  filters: {
    status?: DecisionStatus;
    experiment_id?: number;
    investigation_id?: number;
    release_id?: number;
    due_only?: boolean;
  } = {},
) {
  return useQuery({
    queryKey: ["decisions", filters],
    queryFn: async () => unwrap(await api.GET("/api/decisions", { params: { query: filters } })),
  });
}

export function useDecision(id: number | null) {
  return useQuery({
    queryKey: ["decisions", "one", id],
    queryFn: async () =>
      unwrap(await api.GET("/api/decisions/{decision_id}", { params: { path: { decision_id: id! } } })),
    enabled: id !== null,
  });
}

export function useDecisionMutations() {
  const qc = useQueryClient();
  const invalidate = () => qc.invalidateQueries({ queryKey: ["decisions"] });
  const create = useMutation({
    mutationFn: async (body: DecisionCreate) => unwrap(await api.POST("/api/decisions", { body })),
    onSuccess: invalidate,
  });
  const update = useMutation({
    mutationFn: async (args: { id: number; body: DecisionUpdate }) =>
      unwrap(
        await api.PATCH("/api/decisions/{decision_id}", {
          params: { path: { decision_id: args.id } },
          body: args.body,
        }),
      ),
    onSuccess: invalidate,
  });
  const remove = useMutation({
    mutationFn: async (id: number) =>
      unwrap(await api.DELETE("/api/decisions/{decision_id}", { params: { path: { decision_id: id } } })),
    onSuccess: invalidate,
  });
  return { create, update, remove };
}

// --------------------------------------------------------------------------- search

export function useSearch(q: string, enabled = true) {
  return useQuery({
    queryKey: ["search", q],
    queryFn: async () => unwrap(await api.GET("/api/search", { params: { query: { q, limit: 5 } } })),
    enabled: enabled && q.trim().length >= 2,
    staleTime: 60_000,
    placeholderData: (prev) => prev,
  });
}

File: apps/api/probelens/schemas/ops.py
"""Schemas for releases, SOPs & checklists, knowledge, feedback, decisions, search."""

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from probelens.models.enums import (
    ChecklistStatus,
    DecisionStatus,
    FeedbackSentiment,
    FeedbackSource,
    FeedbackStatus,
    ReleaseStatus,
)
from probelens.schemas.auth import UserSummary

Platform = Literal["android", "ios", "web", "all"]


class EntityRef(BaseModel):
    """Enough to render a link to another object without a second request."""

    type: str
    id: int
    title: str
    status: str | None = None


class ReleaseEventOut(BaseModel):
    id: int
    occurred_at: datetime
    kind: str
    note: str
    actor: UserSummary | None

    model_config = {"from_attributes": True}


class ChecklistItem(BaseModel):
    key: str
    label: str
    owner_role: str | None = None
    done: bool = False
    owner_id: int | None = None
    done_at: datetime | None = None


class ChecklistOut(BaseModel):
    id: int
    sop_id: int | None
    release_id: int | None
    title: str
    owner: UserSummary
    status: ChecklistStatus
    items: list[ChecklistItem]
    done_count: int
    total_count: int
    updated_at: datetime

    model_config = {"from_attributes": True}


class ChecklistProgress(BaseModel):
    done: int
    total: int


class ReleaseSummary(BaseModel):
    id: int
    version: str
    name: str
    platform: str
    status: ReleaseStatus
    release_date: date
    rollout_percent: int
    owner: UserSummary
    affected_areas: list[str]
    experiment_id: int | None
    checklist_progress: ChecklistProgress | None = None
    open_investigations: int = 0
    updated_at: datetime

    model_config = {"from_attributes": True}


class ImpactMetric(BaseModel):
    metric_key: str
    label: str
    format: str
    higher_is_better: bool
    before: float | None
    after: float | None
    abs_change: float | None
    rel_change: float | None
    formatted_before: str
    formatted_after: str
    formatted_change: str
    tone: Literal["good", "bad", "neutral", "unknown"]


class ReleaseImpact(BaseModel):
    before_start: date
    before_end: date
    after_start: date
    after_end: date
    scope: list[str]
    metrics: list[ImpactMetric]
    note: str


class ReleaseOut(ReleaseSummary):
    description: str
    timeline: list[ReleaseEventOut]
    checklists: list[ChecklistOut]
    investigations: list[EntityRef]
    decisions: list[EntityRef]
    experiment: EntityRef | None
    created_at: datetime


class ReleaseCreate(BaseModel):
    version: str = Field(min_length=1, max_length=40)
    name: str = Field(min_length=3, max_length=200)
    description: str = ""
    platform: Platform
    release_date: date
    affected_areas: list[str] = Field(default_factory=list)
    experiment_id: int | None = None
    sop_id: int | None = Field(default=None, description="Run this SOP as the launch checklist")


class ReleaseUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=3, max_length=200)
    description: str | None = None
    status: ReleaseStatus | None = None
    release_date: date | None = None
    rollout_percent: int | None = Field(default=None, ge=0, le=100)
    affected_areas: list[str] | None = None
    experiment_id: int | None = None
    note: str | None = Field(default=None, description="Timeline note explaining the change")


class ReleaseNoteCreate(BaseModel):
    note: str = Field(min_length=1, max_length=2000)
    kind: Literal["note", "link"] = "note"


class SopItem(BaseModel):
    key: str
    label: str
    owner_role: str | None = None


class SopOut(BaseModel):
    id: int
    title: str
    description: str
    category: str
    owner: UserSummary
    items: list[SopItem]
    run_count: int = 0
    updated_at: datetime

    model_config = {"from_attributes": True}


class SopCreate(BaseModel):
    title: str = Field(min_length=3, max_length=200)
    description: str = ""
    category: str = Field(min_length=2, max_length=60)
    items: list[SopItem] = Field(min_length=1)


class SopUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=200)
    description: str | None = None
    category: str | None = Field(default=None, min_length=2, max_length=60)
    items: list[SopItem] | None = None


class ChecklistCreate(BaseModel):
    sop_id: int
    release_id: int | None = None
    title: str | None = None


class ChecklistItemToggle(BaseModel):
    done: bool


class KnowledgeSummary(BaseModel):
    id: int
    title: str
    tags: list[str]
    author: UserSummary
    excerpt: str
    updated_at: datetime


class KnowledgeOut(BaseModel):
    id: int
    title: str
    body: str
    tags: list[str]
    author: UserSummary
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class KnowledgeCreate(BaseModel):
    title: str = Field(min_length=3, max_length=200)
    body: str = Field(min_length=1)
    tags: list[str] = Field(default_factory=list)


class KnowledgeUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=200)
    body: str | None = Field(default=None, min_length=1)
    tags: list[str] | None = None


class StakeholderRef(BaseModel):
    id: int
    name: str
    team: str
    title: str

    model_config = {"from_attributes": True}


class FeedbackOut(BaseModel):
    id: int
    source: FeedbackSource
    theme: str
    body: str
    sentiment: FeedbackSentiment
    status: FeedbackStatus
    platform: str | None
    received_on: date
    stakeholder: StakeholderRef | None
    submitted_by: UserSummary
    linked: EntityRef | None
    created_at: datetime


class FeedbackCreate(BaseModel):
    source: FeedbackSource
    theme: str = Field(min_length=2, max_length=60)
    body: str = Field(min_length=3)
    sentiment: FeedbackSentiment = FeedbackSentiment.neutral
    platform: str | None = None
    received_on: date | None = None
    stakeholder_id: int | None = None
    linked_entity_type: str | None = None
    linked_entity_id: int | None = None


class FeedbackUpdate(BaseModel):
    status: FeedbackStatus | None = None
    theme: str | None = Field(default=None, min_length=2, max_length=60)
    sentiment: FeedbackSentiment | None = None
    linked_entity_type: str | None = None
    linked_entity_id: int | None = None
    unlink: bool | None = None


class ThemeSummary(BaseModel):
    theme: str
    total: int
    open: int
    negative: int
    last_received: date


class DecisionSummary(BaseModel):
    id: int
    title: str
    status: DecisionStatus
    owner: UserSummary
    decided_on: date
    follow_up_date: date | None
    follow_up_due: bool
    linked: list[EntityRef]
    decision: str
    updated_at: datetime


class DecisionOut(DecisionSummary):
    context: str
    evidence: str
    alternatives: str
    expected_impact: str
    investigation_id: int | None
    experiment_id: int | None
    release_id: int | None
    created_at: datetime


class DecisionCreate(BaseModel):
    title: str = Field(min_length=3, max_length=200)
    context: str = ""
    evidence: str = ""
    alternatives: str = ""
    decision: str = Field(min_length=3)
    expected_impact: str = ""
    status: DecisionStatus = DecisionStatus.decided
    decided_on: date | None = None
    follow_up_date: date | None = None
    investigation_id: int | None = None
    experiment_id: int | None = None
    release_id: int | None = None


class DecisionUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=200)
    context: str | None = None
    evidence: str | None = None
    alternatives: str | None = None
    decision: str | None = Field(default=None, min_length=3)
    expected_impact: str | None = None
    status: DecisionStatus | None = None
    decided_on: date | None = None
    follow_up_date: date | None = None
    investigation_id: int | None = None
    experiment_id: int | None = None
    release_id: int | None = None


class SearchHit(BaseModel):
    type: str
    id: int
    title: str
    subtitle: str
    status: str | None = None
    rank: float
    updated_at: datetime | None = None
    extra: dict[str, Any] = Field(default_factory=dict)


class SearchResponse(BaseModel):
    query: str
    total: int
    groups: dict[str, list[SearchHit]]
